### ISSUE TITLE
`foot --maximized` starts with ssd off screen and gap the size of the ssd at the bottom

### ISSUE THREAD

**Issue by Flrian** [labwc/labwc#760]
Started with one of the commits from the last 24h I think.

**Comment by Consolatis** [comment]
Can confirm.
@jlindgren90 seems we missed something somewhere. Thoughts?

**Comment by jlindgren90** [comment]
I'll take a look. I expect it's probably related to one of my changes, not sure which one.

**Comment by jlindgren90** [comment]
Bisecting points to 103a2a8f6c38a8eb664160e369477cad98003598 as the culprit. Still working to understand why.

**Comment by Consolatis** [comment]
```
00:00:06.526 [../src/view.c:504] apply_special
00:00:06.527 [../src/view.c:513] apply_special non-special
00:00:06.527 [../src/view.c:504] apply_special
00:00:06.527 [../src/view.c:506] apply_special maximize
00:00:06.630 [../src/xdg.c:337] First time xdg map, enabling decorations: True
00:00:06.639 [../src/xdg.c:305] Adjusting view by 1,2
```
It seems we are maximizing before the map is actually called. Not sure why

**Comment by Consolatis** [comment]
Right.. `handle_request_maximize()` in xdg.c
```
00:00:02.595 [../src/view.c:504] [0xaaaaab92c300] apply_special
00:00:02.595 [../src/view.c:513] [0xaaaaab92c300] apply_special non-special
00:00:02.595 [../src/xdg.c:172] [0xaaaaab92c300] maximize request
00:00:02.595 [../src/view.c:504] [0xaaaaab92c300] apply_special
00:00:02.595 [../src/view.c:506] [0xaaaaab92c300] apply_special maximize
00:00:02.668 [../src/xdg.c:338] [0xaaaaab92c300] First time xdg map, enabling decorations: True
00:00:02.668 [../src/xdg.c:306] Adjusting view by 1,26
```
Does it need a `view->mapped` check? Or should it create the SSD?

**Comment by jlindgren90** [comment]
> It seems we are maximizing before the map is actually called. Not sure why

Yes. It seems we would maximize first, then center the view in `map()`. 103a2a8f6c38a8eb664160e369477cad98003598 isn't really at fault because it only made the center *actually work*.

The easy fix seems to be add an `if (view->been_mapped)` to `handle_request_maximize()` and `handle_request_fullscreen()`. But in my testing, this causes a visual glitch: the view appears briefly in a non-maximized geometry and is then maximized afterward.

I suspect we could solve this better by following the approach used in `xwayland.c` (related: a417c1c76118321ea85741f95da0a6db833095d9).

**Comment by Consolatis** [comment]
Another workaround would be to delay `wlr_scene_node_set_enabled()` right above `view->mapped = true`

> I suspect we could solve this better by following the approach used in xwayland.c (related: https://github.com/labwc/labwc/commit/a417c1c76118321ea85741f95da0a6db833095d9).

I agree, there isn't really a point in `xdg_position` calling `view_center()` (in the `parent_xdg_toplevel` case) if the view is opened in maximized / fullscreen state.

**Comment by Consolatis** [comment]
This could be a short term fix but I am fine with combining fixing this issue with also keeping natural geometry synced properly.

<details><summary>fix_xdg_maximized_startup.diff</summary>

```diff
diff --git a/src/view.c b/src/view.c
index a42cde2..74b3e01 100644
--- a/src/view.c
+++ b/src/view.c
@@ -669,7 +669,10 @@ view_set_decorations(struct view *view, bool decorations)
 		} else {
 			undecorate(view);
 		}
-		view_apply_special_geometry(view);
+		if (view->mapped) {
+			/* We may need to update maximized / tiled geometry */
+			view_apply_special_geometry(view);
+		}
 	}
 }
 
diff --git a/src/xdg.c b/src/xdg.c
index 8d39a03..a289cda 100644
--- a/src/xdg.c
+++ b/src/xdg.c
@@ -169,6 +169,9 @@ static void
 handle_request_maximize(struct wl_listener *listener, void *data)
 {
 	struct view *view = wl_container_of(listener, view, request_maximize);
+	if (!view->mapped) {
+		return;
+	}
 	view_maximize(view, xdg_toplevel_from_view(view)->requested.maximized,
 		/*store_natural_geometry*/ true);
 }
@@ -177,6 +180,9 @@ static void
 handle_request_fullscreen(struct wl_listener *listener, void *data)
 {
 	struct view *view = wl_container_of(listener, view, request_fullscreen);
+	if (!view->mapped) {
+		return;
+	}
 	struct wlr_xdg_toplevel *xdg_toplevel = xdg_toplevel_from_view(view);
 	view_set_fullscreen(view, xdg_toplevel->requested.fullscreen,
 		xdg_toplevel->requested.fullscreen_output);
@@ -284,7 +290,6 @@ position_xdg_toplevel_view(struct view *view)
 			output_usable_area_from_cursor_coords(view->server);
 		view->current.x = box.x;
 		view->current.y = box.y;
-		view_center(view);
 	} else {
 		/*
 		 * If child-toplevel-views, we center-align relative to their
@@ -350,6 +355,8 @@ xdg_toplevel_view_map(struct view *view)
 		} else if (!view->maximized && requested->maximized) {
 			view_maximize(view, true,
 				/*store_natural_geometry*/ true);
+		} else {
+			view_center(view);
 		}
 
 		view_moved(view);
```

</details>

`foot -F`
```
00:00:07.663 [../src/xdg.c:343]         [0xaaaaab975330] First time xdg map, enabling decorations: True
00:00:07.663 [../src/xdg.c:311]         [0xaaaaab975330] Adjusting view by 1,26
00:00:07.664 [../src/view.c:454]        [0xaaaaab975330] apply_fullscreen
```

`foot -m`
```
00:00:14.045 [../src/xdg.c:343]         [0xaaaaab9797a0] First time xdg map, enabling decorations: True
00:00:14.045 [../src/xdg.c:311]         [0xaaaaab9797a0] Adjusting view by 1,26
00:00:14.045 [../src/xdg.c:360]         [0xaaaaab9797a0] Maximizing on first time xdg map
00:00:14.046 [../src/view.c:505]        [0xaaaaab9797a0] apply_special
00:00:14.046 [../src/view.c:507]        [0xaaaaab9797a0] apply_special maximize
```

**Comment by jlindgren90** [comment]
> We may still need the .x and .y assignments to figure out which output to use for maximizing / fullscreen though.

This is a good point and highlights a related but pre-existing issue: `foot --maximized` always seems to appear on the top-left output, regardless of what output the cursor is on. That happens even before any of my recent changes. I have a fix in my own `master` branch, but unfortunately it depends on a series of changes that aren't ready for merge yet.

**Comment by jlindgren90** [comment]
> This could be a short term fix but I am fine with combining fixing this issue with also keeping natural geometry synced properly.

See #765 for another attempt at a (very) minimal fix. Doesn't address the natural-geometry issue though.

**Comment by Consolatis** [comment]
> `foot --maximized` always seems to appear on the top-left output, regardless of what output the cursor is on.

Is this also the case with the diff from above?
It should actually fix that as a side-effect as well, I assume the reason for that was that the maximize was called before the map which then just used the initial values of view->current.x/y and thus the left-most output. Then the map got called, adjusted x and y but the view was already maximized and thus view_maximize() just returns and does nothing.

**Comment by jlindgren90** [comment]
@Consolatis I haven't tested your exact change tbh, but when I added similar "if mapped" checks, it resulted in visual glitches. That's why I left those checks out of #765 and just focused on restoring the status quo.

So I'm not sure in the long run if it's best to delay the maximize/fullscreen until map. I'd vote for opening a new issue for further discussion.

**Comment by Consolatis** [comment]
tl;dr my proposal sucks and @jlindgren90 knows what he is doing

wayland is asynchronous in regards of sizes but positions are instant.
If we prevent `request_maximize` and `request_fullscreen` for non-mapped views and then have an applcation starting in maximized or fullscreen mode we will see the visual glitches @jlindgren90 was mentioning:
After mapping the window for the first time it will have some size, usually supplied by the application + (in this case) the request to maximize it. We will do so by calculating the new size and sending that to the application (xdg_toplevel_view_configure) and then show the current state. Once the client acknowledges the new (maximized) size and attaches a new buffer the maximized state is working as intended but we still saw the smaller window in between those two steps.

By having the `request_maximize` and `request_fullscreen` calls available to applications even though they are not mapped we can send the correct size of what their buffer should look like which they in turn can then use when requesting being mapped.
Thus, the sizes should match and we won't see any visual glitches like a smaller window for a split second.

So the underlying issue here *might* be that at that point when the application requests to be maximized without being mapped the SSD doesn't exist (because we are only creating it in the map() request handler). Thus the sizes don't match and we are doing something which then causes the offset. I will investigate this some more after I got some sleep, I've been awake for about 30 hours already. A few days more for the bug to exist doesn't really do any harm. I'd really like to understand why #765 fixes the bug and what the actual underlying issue is so we can prevent similar bugs in the future and know what to look for.

**Comment by jlindgren90** [comment]
Get some good sleep @Consolatis :) I'm traveling over the weekend and will be away from a keyboard myself.

**Comment by Consolatis** [comment]
Should be fixed by #765

@Flrian please speak out if it isn't. And thanks for reporting as usual.
